sdk/python/lib/pulumi/dynamic/__main__.py

```diff
@@ -0,0 +1,176 @@
+# limitations under the License.
+
+import asyncio
+import base64
+from concurrent import futures
+import time
+
+import dill
+import grpc
+from google.protobuf import empty_pb2
+from pulumi.runtime import proto, rpc
+from pulumi.runtime.proto import provider_pb2_grpc, ResourceProviderServicer
+from pulumi.dynamic import ResourceProvider
+
+_ONE_DAY_IN_SECONDS = 60 * 60 * 24
+PROVIDER_KEY = "__provider"
+
+def get_provider(props) -> ResourceProvider:
+    byts = base64.b64decode(props[PROVIDER_KEY])
+    return dill.loads(byts)()
+
+class DynamicResourceProviderServicer(ResourceProviderServicer):
+    def CheckConfig(self, request, context):
+        context.set_code(grpc.StatusCode.UNIMPLEMENTED)
+        context.set_details("CheckConfig is not implemented by the dynamic provider")
+        raise NotImplementedError("CheckConfig is not implemented by the dynamic provider")
+
+    def DiffConfig(self, request, context):
+        context.set_code(grpc.StatusCode.UNIMPLEMENTED)
+        context.set_details("DiffConfig is not implemented by the dynamic provider")
+        raise NotImplementedError("DiffConfig is not implemented by the dynamic provider")
+
+    def Invoke(self, request, context):
+        context.set_code(grpc.StatusCode.UNIMPLEMENTED)
+        context.set_details("Invoke is not implemented by the dynamic provider")
+        raise NotImplementedError("unknown function %s" % request.token)
+
+    def Diff(self, request, context):
+        olds = rpc.deserialize_properties(request.olds)
+        news = rpc.deserialize_properties(request.news)
+        if news[PROVIDER_KEY] == rpc.UNKNOWN:
+            provider = get_provider(olds)
+        else:
+            provider = get_provider(news)
+        result = provider.diff(request.id, olds, news)
+        fields = {}
+        if result.changes is not None:
+            if result.changes:
+                fields["changes"] = proto.DiffResponse.DIFF_SOME # pylint: disable=no-member
+            else:
+                fields["changes"] = proto.DiffResponse.DIFF_NONE # pylint: disable=no-member
+        else:
+            fields["changes"] = proto.DiffResponse.DIFF_UNKNOWN # pylint: disable=no-member
+        if result.replaces is not None:
+            fields["replaces"] = result.replaces
+        if result.delete_before_replace is not None:
+            fields["deleteBeforeReplace"] = result.delete_before_replace
+        return proto.DiffResponse(**fields)
+
+    def Update(self, request, context):
+        olds = rpc.deserialize_properties(request.olds)
+        news = rpc.deserialize_properties(request.news)
+        provider = get_provider(news)
+
+        result = provider.update(request.id, olds, news)
+        outs = {}
+        if result.outs is not None:
+            outs = result.outs
+        outs[PROVIDER_KEY] = news[PROVIDER_KEY]
+
+        loop = asyncio.new_event_loop()
+        outs_proto = loop.run_until_complete(rpc.serialize_properties(outs, {}))
+        loop.close()
```

Comment:
consider extracting these lines.  i see this duplicated a lot in this type.

Reply:
We already use this exact pattern in a half dozen other places in all the rest of our Python RPC serialization logic.  I'd prefer to leave this consistent for now.